notes/fomega/evaluation/builtin-evaluation.md

```diff
@@ -0,0 +1,347 @@
+# Efficiently evaluating builtin functions in the CK and CEK machines
+
+(This is a bit impressionistic: it's describing ongoing work and discussions.)
+
+It's important that builtin evaluation should be efficient in our
+Plutus Core evaluators.  There's some machinery known as the constant
+application machinery (CAM for short) which takes the name of a
+builtin and a list of arguments of some unspecified type `term` which
+is required to provide some `Constant`-like constructor which contains
+values of built-in types.  `Constant` terms can be converted back into
+Haskell values which can be used by the Haskell implementation of a
+builtin. Non-`Constant` terms are opaque to builtins and can be fed to
+them as arguments, but pass through undigested, as it were.  This
+mechanism is described in more detail
+[here](../../../language-plutus-core/docs/Builtins.md) and
+[here](../../../language-plutus-core/docs/Constant-application.md).
+
+The question considered here is how the evaluators should interact
+with the CAM, especially when we're dealing with unsaturated builtins.
+The basic problem is how to tell if an application has the required
+number of arguments.  In the original version of the CEK machine, this
+was handled entirely by the CAM.
+
+### CEK strategy
+
+Recall that the CEK machine uses a stack of _frames_ containing
+information about the context surrounding the term currently being
+evaluated.  When an application `[M N]` is encountered, firstly `M` is
+reduced to a value (`F` say), which is saved in a frame on the stack,
+then `N` is reduced to a value `V`, then `F` is applied to `V`: this
+is the step that concerns us.
+
+There are three possibilities:
+
+ 1. `F` is a lambda term of the form `(lam x t body)`.  In this case,
+ evaluation proceeds in the CEK machine by binding `V` to `x` and then
+ proceeding to evaluate `body`. (In the CK machine, `V` is substituted
+ for every occurrence of `x` in `body`, which is then evaluated).
+
+ 2. `F` is an application of a builtin function to a (possibly empty) list of values.
+
+ 3. `F` is something else (a constant for example): in this case an error occurs.
+
+Step 2 is the one that has to be efficient.  In the original CEK machine, a function
+`termAsPrimIterApp` is called.  This takes a term and checks whether it is a sequence of
+type instantiations and applications with the name of a builtin function at the head, for example
+
+```
+  [{([builtin b) V₁ V₂] T₁} V₃]
```

Comment:
There seems to be an interleaving of values and types here that I wasn't expecting. I thought this was describing the existing machinery that doesn't allow such interleaving.

(and some mismatched brackets)

Reply:
Ah, that's mostly what we're arguing about.  The question is whether we should allow interleaving or not, and the opinion seems to be that we should.  The problem then is what we erase type application to.  We erase `abs` to `delay` and `{}`to `force`, and then we have to ask what happens to instantiations of partially-applied builtins.  A builtin can't compute at all until it gets its final argument, so the only case where there might be a problem is if we have an instantiation after the final term argument.  It's not clear (to me, anyway) if we'd ever want that case though.

Reply:
I think returning something of type `forall a . t` from a builtin should be handled by the ordinary erasure machinery as the builtin is no-longer in the picture. But, I could be wrong.

Reply:
That's the thing I'm not sure about either.  Is there any difference between a builtin which consumes all of its term arguments but doesn't compute until it gets instantiated and one which consumes its arguments and returns an abstraction which you can then instantiate?  The instantiation definitely has to be handled by the evaluator because we're not actually passing types to the CAM (and if we did it would probably be unpleasant: we'd have to pass elements of a sum type and the machinery would have to discriminate on those, which would be inefficient.  Maybe we could pass units, but then how would you tell them apart from units passed as term arguments?  Would you need to?)

Reply:
Ah, I guess was expecting that the CAM would have to deal with this somehow. Perhaps it would help if we tried to write down some typing rules and either a reduction or CK semantics for this.